src/docker/ContextManager.ts

```diff
@@ -0,0 +1,201 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ExecOptions } from 'child_process';
+import * as fs from 'fs';
+import * as fse from 'fs-extra';
+import * as os from 'os';
+import * as path from 'path';
+import { URL } from 'url';
+import { Event, EventEmitter, workspace } from 'vscode';
+import { Disposable } from 'vscode';
+import { callWithTelemetryAndErrorHandling, IActionContext } from 'vscode-azureextensionui';
+import { LineSplitter } from '../debugging/coreclr/lineSplitter';
+import { ext } from '../extensionVariables';
+import { AsyncLazy } from '../utils/lazy';
+import { execAsync, spawnAsync } from '../utils/spawnAsync';
+import { DockerContext, DockerContextInspection } from './Contexts';
+import { DockerodeApiClient } from './DockerodeApiClient/DockerodeApiClient';
+import { DockerServeClient } from './DockerServeClient/DockerServeClient';
+
+// CONSIDER
+// Any of the commands related to Docker context can take a very long time to execute (a minute or longer)
+// if the current context refers to a remote Docker engine that is unreachable (e.g. machine is shut down).
+// Consider having our own timeout for execution of any context-related Docker CLI commands.
+// The following timeout is for _starting_ the command only; in the current implementation there is no timeot
+// for command duration.
+const ContextCmdExecOptions: ExecOptions = { timeout: 5000 }
+
+const dockerConfigFile = path.join(os.homedir(), '.docker', 'config.json');
+const dockerContextsFolder = path.join(os.homedir(), '.docker', 'contexts', 'meta');
+
+const WindowsLocalPipe = 'npipe:////./pipe/docker_engine';
+const UnixLocalPipe = 'unix:///var/run/docker.sock';
+
+const defaultContext: Partial<DockerContext> = {
+    Name: 'default',
+    Description: 'Current DOCKER_HOST based configuration',
+    CreatedTime: undefined, // Not defined for contexts

```

Comment:
If the type of the variable is `Partial<DockerContext>` do you need to set the `CreatedTime` to undefined?